### ISSUE TITLE
KubernetesReconcilerProcessor incorrectly instantiate beans

### ISSUE THREAD

**Issue by asavov** [kubernetes-client/java#1511]
`KubernetesReconcilerProcessor` is `BeanFactoryPostProcessor`. And `BeanFactoryPostProcessor` doc explicitly states that:

_A BeanFactoryPostProcessor may interact with and modify **bean definitions**, but **never** bean instances._

Still `KubernetesReconcilerProcessor` does not honour that:

```
  @Override
  public void postProcessBeanFactory(ConfigurableListableBeanFactory beanFactory)
      throws BeansException {
    String[] names = beanFactory.getBeanNamesForType(Reconciler.class);
    for (String name : names) {
      // THIS IS VIOLATION
      Reconciler reconciler = (Reconciler) beanFactory.getBean(name);
```
This results in error while instantiating my Reconciler.

```
@KubernetesReconciler(
        watches = @KubernetesReconcilerWatches({
                @KubernetesReconcilerWatch(apiTypeClass = V1Pod.class)
        }))
@Component
public class PodController implements Reconciler {
    @Autowired
    public PodController(SharedInformer<V1Pod> informer, Lister<V1Pod> lister) {}
```
Error:
`Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [PodController]: No default constructor found; nested exception is java.lang.NoSuchMethodException: PodController.<init>()`

**Comment by brendandburns** [comment]
cc @yue9944882 @dsyer for Spring.

**Comment by asavov** [comment]
@yue9944882 @dsyer I'm really open to work on that If you share some guidelines/hints/recommendation what's the **right** approach (the Spring way) to configure and instantiate k8s Controllers/Reconcilers. We might discuss this over zoom.

**Comment by dsyer** [comment]
`KubernetesReconcilerProcessor` should be able to create and register `BeanDefinitions` without actually instantiating anything (no need to call `getBean()` or `registerSingleton()`). It only uses the reconciler to find out its type, and that information should be available in the `BeanDefinition`. It's a pretty simple class, so that change ought to be easy. There is one integration test.

OTOH I don't find the annotations all that useful, and I'd actually rather just throw them away and ditch the `KubernetesReconcilerProcessor` longer term. But we don't have to do that right now, since the changes to fix this issue would be minor.

**Comment by brendandburns** [comment]
fwiw, all of the Spring stuff is pretty experimental/early. I don't think there is a "right" way yet, just some early exploration.

@asavov do you have thoughts about what you as a Spring user would expect?

**Comment by asavov** [comment]
To have core Spring capabilities (such as @Autowire, @Component, etc) for `MyController implements Reconciler` which with current impl is not functional at all.

**Comment by dsyer** [comment]
So do you think you might have time to fix it? As I said above I don’t think it’s a big job.

**Comment by asavov** [comment]
@dsyer I spent some time on providing a fix: https://github.com/kubernetes-client/java/pull/1516

This is my first contribution so need to clear all the steps. Still the code is reviewable.

**Comment by dsyer** [comment]
This closed by #1516 I think.
